internal/charmstore/store.go

```diff
@@ -788,3 +797,44 @@ func (s *Store) SynchroniseElasticsearch() error {
 	}
 	return nil
 }
+
+//GetPromulgatedURL finds the promulgatedURL that should be used for
+//this newly uploaded charm, if the charm should be promulgated,
+//othewise it returns nil. An error is returned if there is a problem
+//communicating with the storage.
+func (s *Store) GetPromulgatedURL(id *charm.Reference) (*charm.Reference, error) {
+	var baseEntity mongodoc.BaseEntity
+	err := s.DB.BaseEntities().FindId(baseURL(id)).One(&baseEntity)
+	if err != nil {
+		if err == mgo.ErrNotFound {
+			// If there is no base entity then by definition it cannot be promulgated.
+			return nil, nil
+		}
+		return nil, errgo.Mask(err)
+	}
+	if !baseEntity.Promulgated {
+		return nil, nil
+	}
+	var entity mongodoc.Entity
+	err = s.DB.Entities().Find(bson.D{{"name", id.Name}, {"series", id.Series}, {"promulgated-url", bson.D{{"$exists", true}}}}).Sort("-promulgated-revision").One(&entity)
```

Comment:
Perhaps select the only fields that are required?

Also, I think I'd find this line easier to read if it was broken up a little.